File: Solutions/572-SubtreeOfAnotherTree/SubtreeOfAnotherTree.ts
import { TreeNode } from '../../types';

function isSubtree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
    if (!subRoot) return true;
    if (!root) return false;

    if (sameTree(root, subRoot)) return true;
    return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);

    function sameTree(root: TreeNode | null, sub: TreeNode | null): boolean {
        if (!root && !sub) return true;
        if (root && sub && root.val === sub.val) {
            return (
                sameTree(root.left, sub.left) && sameTree(root.right, sub.right)
            );
        }
        return false;
    }
}

File: Solutions/572-SubtreeOfAnotherTree/SubtreeOfAnotherTree.kt
class Solution {
    fun isSubtree(root: TreeNode?, subRoot: TreeNode?): Boolean {
        if(subRoot == null) return true;
        if(root == null) return false;

        if(root.`val` == subRoot.`val` && isSameTree(root, subRoot)){
            return true
        } else {
            return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot)
        }
    }

    fun isSameTree(rootOne: TreeNode?, rootTwo: TreeNode?): Boolean {
        if(rootOne == null && rootTwo == null) return true;
        if(
            rootOne == null ||
            rootTwo == null ||
            rootOne.`val` != rootTwo.`val`
        ) return false;

        return isSameTree(rootOne.left, rootTwo.left) && isSameTree(rootOne.right, rootTwo.right)
    }
}

